### PR TITLE
Dev branch

### PR DESCRIPTION
**Pull request by IBUNHABIBU**

### CODE REVIEW

File: Enumerables.rb [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        i = 0
+        while i<self.size
+            yield self[i]
+            i+=1
+        end
+    end
+
+    def my_each_with_indeitem
+        i = 0
+        result = []
```
Comment (mosaaleb) [inline Enumerables.rb]: **[OPTIONAL]** The default method of ruby is `each_with_index`. For better readability let's rename it to `my_each_with_index`

File: Enumerables.rb [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    end
+
+    def my_any?
+        my_each{|item| return true if yield item}
+        false
+    end
+
+    def my_none
+        my_each{|item| return false if yield item}
+        true
```
Comment (mosaaleb) [inline Enumerables.rb]: - **[OPTIONAL]** It's a common convention in ruby that any methods return `true` or `false` to end with question mark `?`

File: Enumerables.rb (hunk 1 above)
Comment (mosaaleb) [inline Enumerables.rb]: [OPTIONAL] The default method of ruby is each_with_index. For better readability let's rename it to my_each_with_index.

File: Enumerables.rb (hunk 2 above)
Comment (mosaaleb) [inline Enumerables.rb]: **[OPTIONAL]** It's a common convention in ruby that any methods return true or false to end with question mark ?

File: Enumerables.rb [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        result
+    end
+end
+weekdays=["monday","tuesday","wednesday","thursday"]
+weekdays.my_each{|day| puts day.upcase}
+weekdays.my_each_with_indeitem{|day,item| puts "#{item}.#{day.upcase}"}
+
+    def multiply_els(array)
+        array.my_inject(4){|item,n| item*n}
+    end
```
Comment (mosaaleb) [inline Enumerables.rb]: - [ ] `multiply_els` method always return zero. As it depends on `my_inject`, `my_inject` logic is not right.

**Review by mosaaleb — CHANGES_REQUESTED** [review]
## Well done @IBUNHABIBU :clap:

Most of the functionality are working great, yet there are some requirement needs work before moving on.

- [ ] Let's fix all stickler errors (identation, spaces, ruby guideliens, ...)
- [ ] Your logic for `my_inject` is not right. You start by setting the result to zero and if the operation is multiplication it will always return `0`. You need to set it to the first element of the array yield it. Don't forget to exclude the first item when applying the block. Also, let's test it on some data before submitting a new review.

Also please see my comments below

Please handle these requirements and submit a new code review.

#### Happy Coding :computer:

File: Enumerables.rb [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        count
+    end
+
+    def my_map(&block)
+        result = []
+        my_each{|item| result<<block.call(item)}
+        result
+    end
+
+    def my_inject(init=nil)
```
Comment (mosaaleb) [inline Enumerables.rb]: - [ ] `my_inject` will always return zero if applied a multiplication operation on it as the initial value of result is zero so multipled by any other number will be always zero.

**Comment by IBUNHABIBU** [comment]
The check says that queed 5hrs ago
what cause this?

File: enumerable_methods.rb [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    result = []
+    my_each { |item| result << block.call(item) }
+    result
+  end
+
+  def my_inject(_init = nil)
+    result = 0
+    my_each { |item| result = yield(result, item) }
+    result
+  end
```
Comment (maelfosso) [inline enumerable_methods.rb:77]: There is still some problems
```ruby
irb(main):004:0> [5,6,7,8,9,10].inject { |sum, n| sum + n }
=> 45
irb(main):005:0> [5,6,7,8,9,10].my_inject { |sum, n| sum + n }
=> 45
irb(main):006:0> [5,6,7,8,9,10].inject(5) { |sum, n| sum + n }
=> 50
irb(main):007:0> [5,6,7,8,9,10].my_inject(5) { |sum, n| sum + n }
=> 45
irb(main):008:0> [5,6,7,8,9,10].inject { |prod, n| prod * n }
=> 151200
irb(main):009:0> [5,6,7,8,9,10].my_inject { |prod, n| prod * n }
=> 0
```

File: enumerable_methods.rb [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+  def my_select
+    result = []
+    my_each { |item| result << item if yield item }
+    result
+  end
+
+  def my_all?
+    my_each { |item| return false unless yield item }
+    true
+  end
```
Comment (maelfosso) [inline enumerable_methods.rb:40]: It's not working when **`no block given`**
```ruby
irb(main):010:0> [nil, true, 99].all?
=> false
irb(main):011:0> [nil, true, 99].my_all?
Traceback (most recent call last):
        5: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        4: from (irb):11
        3: from enumerable_methods.rb:28:in `my_all?'
        2: from enumerable_methods.rb:6:in `my_each'
        1: from enumerable_methods.rb:28:in `block in my_all?'
LocalJumpError (no block given (yield))
```

File: enumerable_methods.rb [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+  def my_any?
+    my_each { |item| return true if yield item }
+    false
+  end
```
Comment (maelfosso) [inline enumerable_methods.rb:45]: Sorry :disappointed:  but it's not working when **`no block given`**
```ruby
irb(main):022:0> [nil, true, 99].any?  
=> true
irb(main):023:0> [nil, true, 99].my_any?  
Traceback (most recent call last):
        5: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        4: from (irb):23
        3: from enumerable_methods.rb:33:in `my_any?'
        2: from enumerable_methods.rb:6:in `my_each'
        1: from enumerable_methods.rb:33:in `block in my_any?'
LocalJumpError (no block given (yield))
```

File: enumerable_methods.rb [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+  def my_none
+    my_each { |item| return false if yield item }
+    true
+  end
```
Comment (maelfosso) [inline enumerable_methods.rb:56]: Not working when **`no block given`**
```ruby
irb(main):040:0> [].none? 
=> true
irb(main):041:0> [].my_none
=> true
irb(main):042:0> [nil].my_none 
Traceback (most recent call last):
        5: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        4: from (irb):42
        3: from enumerable_methods.rb:38:in `my_none'
        2: from enumerable_methods.rb:6:in `my_each'
        1: from enumerable_methods.rb:38:in `block in my_none'
LocalJumpError (no block given (yield))
irb(main):043:0> [nil, false, true].my_none  
Traceback (most recent call last):
        5: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        4: from (irb):43
        3: from enumerable_methods.rb:38:in `my_none'
        2: from enumerable_methods.rb:6:in `my_each'
        1: from enumerable_methods.rb:38:in `block in my_none'
LocalJumpError (no block given (yield))
irb(main):044:0> 
```

**Review by maelfosso — CHANGES_REQUESTED** [review]
Hello @IBUNHABIBU 

# Really nice improvement :clap: 

It's almost done. There is some changes to do.

- Please, indent your code. It's very very important to have a clear code. Most of the sticklers errors could be easily solved from a clear code.
Look at this prompt 
```ruby
$ ruby -wc enumerable_methods.rb 
enumerable_methods.rb:8: warning: mismatched indentations at 'end' with 'while' at 5
Syntax OK
```
That means that even the syntax is OK, there is some indentation problems.

- **`REQUIRED`** Solve all the sticklers errors.

- **`REQUIRED`** There are still some methods which are not working. Kindly look below.

Please, make the corresponding changes.
Don't hesitate to message me on Slack if you have any questions.

File: Enumerables.rb (hunk 1 above)
Comment (IBUNHABIBU) [reply Enumerables.rb]: I think this is autocompletion mistake I was not intending to name like that

**Comment by IBUNHABIBU** [comment]
```Check failure on line 47 in enumerable_methods.rb```

``@stickler-ci
stickler-ci
/ stickler-ci``

```enumerable_methods.rb#L47
C: Metrics/AbcSize: Assignment Branch Condition size for my_count is too high. [16.55/15]
C: Metrics/CyclomaticComplexity: Cyclomatic complexity for my_count is too high. [9/6]
C: Metrics/PerceivedComplexity: Perceived complexity for my_count is too high. [9/7]```
```

I have no idea to correct those errors

File: enumerable_methods.rb (hunk 8 above)
Comment (maelfosso) [reply enumerable_methods.rb:56]: Sorry :disappointed: but it's still not working
```ruby
irb(main):005:0> [nil, false, true].my_none?
=> true
irb(main):006:0> [nil, false, true].none?
=> false
```

File: enumerable_methods.rb (hunk 5 above)
Comment (maelfosso) [reply enumerable_methods.rb:77]: Sorry :disappointed: but it's not working
```ruby
irb(main):009:0> [5,6,7,8,9,10].inject { |prod, n| prod * n }
=> 151200
irb(main):010:0> [5,6,7,8,9,10].my_inject { |prod, n| prod * n }
=> 0
```

File: enumerable_methods.rb (hunk 6 above)
Comment (maelfosso) [reply enumerable_methods.rb:40]: :clap: you fix the bug highlighted up but there is a new bug
```ruby
irb(main):013:0> %w[ant bear cat].all? { |word| word.length >= 3 }
=> true
irb(main):014:0> %w[ant bear cat].my_all? { |word| word.length >= 3 }
=> nil
```

**Review by maelfosso — CHANGES_REQUESTED** [review]
Hello

# :clap: Nice improvement :smile: 

Please there are two bugs you need to fix. Take a look below to see where the problems are.
To understand the stickler errors, please follow those links
**Assignment Branch Condition size for my_count is too high**
- http://redgreenrepeat.com/2017/01/20/understanding-assignment-branch-condition/
- https://stackoverflow.com/questions/55422300/assignment-branch-condition-size-is-too-high
- https://stackoverflow.com/questions/30932732/what-is-meant-by-assignment-branch-condition-size-too-high-and-how-to-fix-it
**Cyclomatic complexity for my_count is too high**
- https://stackoverflow.com/questions/47336195/cyclomatic-complexity-for-some-method-is-too-high
- https://codereview.stackexchange.com/questions/159189/cyclomatic-complexity-for-method-is-too-high

Kindly make the corresponding changes and re-submit a new code review request.
Thanks

File: enumerable_methods.rb (hunk 8 above)
Comment (maelfosso) [reply enumerable_methods.rb:56]: Sorry @IBUNHABIBU there is still a bug here
```ruby
irb(main):004:0> %w{ant bear cat}.none? { |word| word.length == 5 }
=> true
irb(main):005:0> %w{ant bear cat}.my_none? { |word| word.length == 5 }
=> false
```

File: enumerable_methods.rb (hunk 5 above)
Comment (maelfosso) [reply enumerable_methods.rb:77]: Sorry :disappointed: same bugs, it should be fixed
```ruby
irb(main):016:0> [5,6,7,8,9,10].inject { |prod, n| prod * n }
=> 151200
irb(main):017:0> [5,6,7,8,9,10].my_inject { |prod, n| prod * n }
=> 0
```

**Review by maelfosso — CHANGES_REQUESTED** [review]
Hello @IBUNHABIBU ,

:clap: 
:clap: 
# Nice improvements :smile: 

You fixed stickler errors and also most of the existing bugs.
You really are at one step of the last line, don't give up.
Take a look on the comment, there is the method **my_none** which still have a bug.

Please, fix the bugs and submit a new code review request.
Thanks.

File: enumerable_methods.rb [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+    length
+  end
+
+  def my_map(&block)
+    result = []
+    my_each { |item| result << block.call(item) }
+    result
+  end
+
+  def my_inject(init = 0)
```
Comment (bolah2009) [inline enumerable_methods.rb]: #### `REQUIRED`

:bug: As @maelfosso pointed out in[ this comment](#discussion_r322010779), let's fix this bug.

```
irb(main):016:0> [5,6,7,8,9,10].inject { |prod, n| prod * n }
=> 151200
irb(main):017:0> [5,6,7,8,9,10].my_inject { |prod, n| prod * n }
=> 0
```

File: .rubocop.yml [hunk 10]
```diff
@@ -0,0 +1,11 @@
...
+Metrics/LineLength:
+  Max: 120
+
+# Too short methods lead to extraction of single-use methods, which can make
+# the code easier to read (by naming things), but can also clutter the class
+Metrics/MethodLength: 
+  Max: 20
+
+Metrics/BlockLength:
+  ExcludedMethods: ['describe']
```
Comment (bolah2009) [inline .rubocop.yml]: Let's use the appropriate linter file. [Check here](https://github.com/microverseinc/linters-config/blob/master/ruby/.rubocop.yml)

```yml
AllCops:
  Exclude:
    - "bin/*"
    - "db/*"
    - "config/*"
    - "Guardfile"
    - "Rakefile"
  DisplayCopNames: true

Metrics/LineLength:
  Max: 120
Metrics/MethodLength:
  Include: ["app/controllers/*"]
  Max: 20
Metrics/AbcSize:
  Include: ["app/controllers/*"]
  Max: 30
Metrics/ClassLength:
  Max: 150
Metrics/BlockLength:
  ExcludedMethods: ['describe']

Style/Documentation:
  Enabled: false
Style/ClassAndModuleChildren:
  Enabled: false

Layout/AlignHash:
  EnforcedColonStyle: key
Layout/ExtraSpacing:
  AllowForAlignment: false
Layout/MultilineMethodCallIndentation:
  Enabled: true
  EnforcedStyle: indented
```

File: README.md [hunk 11]
```diff
@@ -1 +1,8 @@
 # Enumerables
+## This is the project to implement user defined Enumerable methods
+##### using 
+* ruby
+###  Author 
+Salum Habibu https://github.com/IBUNHABIBU
+
+[ Live project ](https://repl.it/@IBUNHABIBU/Enumerables)
```
Comment (bolah2009) [inline README.md:19]: #### `HINT`

Add a descriptive readme. Details like assignment link (from Odin), code snippet of what the methods can do and what was learned during the project can make the project more descriptive.

File: enumerable_methods.rb [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+end
+
+puts multiply_els([2, 3, 4])
+
+puts [2, 3, 4].inject(0) { |sum, item| sum + item }
+
+puts [].my_none?                                           #=> true
+puts [nil].my_none?                                        #=> true
+puts [nil, false].my_none?                                 #=> true
+puts [nil, false, true].my_none?                           #=> false
```
Comment (bolah2009) [inline enumerable_methods.rb]: #### `HINT`

Let's move all these code examples to the `README` as `code snippet`. :thinking:

**Review by bolah2009 — CHANGES_REQUESTED** [review]
# Hi @IBUNHABIBU  :wave: 

## Awesome Job :clap: and nice improvements from the last review :smile: 

A few things to make before a :+1: 
- Add appropriate linter file (`.rubocop.yml`) and fix linter warning.
  With `rubocop -a` you can fix almost all of the errors, if you have the `rubocop` gem installed: (`gem install rubocop`). :wink: 
- Let's make the suggestions raised by @maelfosso work. :pray: 
  - `my_inject` - [See comment ](#discussion_r322010779) for details.
   #### :bulb: [HINT] Check out [ruby-doc](https://ruby-doc.org/core-2.6.3/Enumerable.html) for expected behaviours of these methods.
- Update PR Description
   update-PR-description

#### :recycle:  Make these changes and submit another code review request. :+1: 
#### Feel free to reach me on slack if you having issues implementing these features. :+1: 

Good luck with your next submission. :star2: 

Cheers
[Bola](https://bolabuari.com/)

**Comment by IBUNHABIBU** [comment]
this error is difficult to remove I tried a lot 
C: Style/IfInsideElse: Convert `if` nested inside `else` to `elsif`.

**Pull request merged**